[Flang][OpenACC][OpenMP] Always add extent to acc/omp.BoundsOp

agozillon opened the pull request:
This patch changes the bounds generation code shared between OpenMP and OpenACC to always attach an extent to the bounds generation. This is currently required for OpenMP descriptor lowering but may not necessarily be required in the case of OpenACC.

agozillon commented:
This is the segment of changes to the bounds from the larger allocatable/pointer patch where the extent is used in the lowering @razvanlupusoru thought it seemed reasonable, however, if we wish to only have this switched on for OpenMP (or toggleable via a passed in argument) we can also do that!

razvanlupusoru reviewed (approved):
Seems reasonable to me. Thank you!

clementval commented on flang/test/Lower/OpenMP/FIR/array-bounds.f90:
```diff
@@ -57,11 +61,12 @@ end subroutine assumed_shape_array
 !ALL: %[[C1:.*]] = arith.constant 1 : index
 !ALL: %[[C2:.*]] = arith.constant 4 : index
-!ALL: %[[BOUNDS:.*]]  = omp.bounds   lower_bound(%[[C1]] : index) upper_bound(%[[C2]] : index) stride(%[[C0]] : index) start_idx(%[[C0]] : index)
+!ALL: %[[BOUNDS:.*]]  = omp.bounds   lower_bound(%[[C1]] : index) upper_bound(%[[C2]] : index) extent(%[[UNDEF]] : index) stride(%[[C0]] : index) start_idx(%[[C0]] : index)
```
This seems risky to have undef in the extent. Wouldn't it be better to not generate the extent if it is coming from a `fir.undefined`?

clementval reviewed (commented):
Just have a question about the undef case.

clementval reviewed (changes requested):
Undef extent needs to be set to zero to be consistent with other code path.

agozillon commented:
@clementval @razvanlupusoru thank you both for the review, I'll update the PR tomorrow with the requested changes!

agozillon committed:
[Flang][OpenACC][OpenMP] Always add extent to acc/omp.BoundsOp

This patch changes the bounds generation code shared between OpenMP
and OpenACC to always attach an extent to the bounds generation. This
is currently required for OpenMP descriptor lowering but may not
neccessarily be required in the case of OpenACC.

clementval commented on flang/test/Lower/OpenMP/array-bounds.f90:
```diff
@@ -64,14 +65,14 @@ end subroutine assumed_shape_array
 !HOST: %[[C1:.*]] = arith.constant 1 : index
 !HOST: %[[C2:.*]] = arith.constant 4 : index
-!HOST: %[[BOUNDS:.*]]  = omp.bounds   lower_bound(%[[C1]] : index) upper_bound(%[[C2]] : index) stride(%[[C0]] : index) start_idx(%[[C0]] : index)
+!HOST: %[[BOUNDS:.*]]  = omp.bounds   lower_bound(%[[C1]] : index) upper_bound(%[[C2]] : index) extent(%[[EXT]] : index) stride(%[[C0]] : index) start_idx(%[[C0]] : index)
```
There seems to be an error here. The extent of `arr_read_write(2:5)` is computable since the lb and ub are constant. 
In the OpenACC flow we never have a 0 extent for a section defined with two constant. The 0 extent comes from arrays defined as `a(1:*)`.

agozillon commented:
Hopefully the last adjustment covers your concerns. Wasn't too sure if the additional plus one was required, I believe it is from what I can gather, but I am perhaps incorrect.

clementval commented on flang/lib/Lower/DirectivesCommon.h:
```diff
@@ -819,6 +819,13 @@ genBoundsOps(fir::FirOpBuilder &builder, mlir::Location loc,
         if (mlir::isa<fir::UndefOp>(extent.getDefiningOp())) {
           extent = zero;
+          if (ubound && lbound &&
+              mlir::isa<mlir::arith::ConstantOp>(ubound.getDefiningOp()) &&
+              mlir::isa<mlir::arith::ConstantOp>(lbound.getDefiningOp())) {
```
Thanks for the fix. You can remove the check for constant. If lb and ub are set we can compute the extent even if those are variables. `a(n:m)` should also have a proper extent compute and not zero. The +1 is correct.

agozillon replied:
Sounds good, I'll remove the constant check!

clementval reviewed (approved):
LGTM. Thanks for the changes
